Implement plugin for Health and Fitness room

panpawn opened the pull request:
Approved by the ROs

HoeenCoder commented on chat-plugins/health-fitness.js:
```diff
@@ -0,0 +1,151 @@
+		this.hfData.quote = quote;
+		this.saveData();
+	}
+	getData() {
```
This could be setup like
```javascript
get data() {
return this.hfData;
}
```
So that `HealthFitness#data` dosent need to be called as a function, it can just be called like a variable.

Zarel replied:
Ugh no do not do this ever

Zarel replied:
Getters and setters should only be used when there's a specific reason why pretending they're variables is a good idea.

Zarel replied:
Although, like, why can't we read `.hfData` directly? Is there a need for this sort of indirection?

panpawn replied:
Good point @Zarel

HoeenCoder commented on chat-plugins/health-fitness.js:
```diff
@@ -0,0 +1,151 @@
+	setQuote(quote) {
+		this.hfData.quote = quote;
+		this.saveData();
+	}
```
Possibly merge these three into one and add an extra argument to pick which piece of data to edit.

Zarel replied:
Disagree.

panpawn commented:
Updated

Morfent commented on chat-plugins/health-fitness.js:
```diff
@@ -0,0 +1,147 @@
+			cardio: '',
+			gym: '',
+			quote: '',
+			__proto__: null,
```
The constructor's not going to be optimizable using `__proto__` in an object literal like this. `Object.create` would probably be a better idea instead, even though this is only getting called once between restarts

panpawn replied:
So, something like this?

```/Javascript
this.hfData = Object.create({
     cardio: '',
     gym: '',
     quote: '',
     __proto__: null,
});
```

Morfent replied:
.No, since that's still using `__proto__`. `Object.assign(Object.create(null), {cardoi: '', gym: '', quote: ''})` allows you to create an object with both a null prototype and properties without needing to mutate `[[Prototype]]`

panpawn replied:
Thanks!

Zarel replied:
Wait, how does the old code mutate `[[Prototype]]`?

Morfent replied:
...you're right actually, I was thinking of when it's used as a setter, not when creating an object.

panpawn replied:
Er, so how should this be done?

Zarel replied:
Your old code was fine. Morfent tends to misunderstand how JavaScript works sometimes. :p Change it back.

panpawn replied:
Okay, I changed it back

panpawn commented:
Updated

panpawn commented:
How does this look now?

sparkychildcharlie commented on chat-plugins/health-fitness.js:
```diff
@@ -0,0 +1,148 @@
+			const gym = HF.hfData.gym || '(not set yet)';
+			const quote = HF.hfData.quote || '(not set yet)';
+			return this.sendReplyBox(
+				Chat.html`<strong>Cardio Challenge:</strong> ${cardio}<br />` +
```
Maybe use Chat.parseText for some of these so links are autolinked?

panpawn replied:
I'm not sure that the room is planning on including links in these. I'll confirm this later :p

Hawkiee replied:
I've used links in the challenges before, usually to explain a more obscure exercise not many people know

Zarel commented on chat-plugins/health-fitness.js:
```diff
@@ -0,0 +1,148 @@
+			cardio: '',
+			gym: '',
+			quote: '',
+			__proto__: null,
```
Unrelatedly: Why is this null-proto at all, if it's loaded from a JSON file?

Zarel replied:
bump

panpawn replied:
I've been busy with school, sorry. Uhh, I didn't realize that an object loaded from a JSON file wasn't susceptible to constructor crashes.

If this isn't the case, I'll update this later today.

Zarel replied:
Objects loaded from JSON files are the same as any other object.

The issue is that you can't load null-prototype objects from JSON files. Which is why I instead use `hasOwnProperty` for them.

panpawn commented:
Updated

Zarel commented:
Is this done? It looks pretty close to me.

AustinXII commented on chat-plugins/health-fitness.js:
```diff
@@ -0,0 +1,147 @@
+		},
+	},
+	healthfitnesshelp: [
+		"/healthfitness - Shows the daily cardio challenge, gym challenge, and quote of the day",
```
this day is missing a period after it. the lines that include "or above" also appear to be missing a period unless that's intentional.

panpawn commented:
Updated:
- Entries now use ``Chat.parseText`` which will allow for PS formatting
- Fixed a missing period in the help command
- Run entries thru the chat filter (by passing thing thru ``this.canTalk(target)`` before setting them for security sake)
- Changed data file path

panpawn commented:
bump?

Zarel commented on chat-plugins/health-fitness.js:
```diff
@@ -0,0 +1,168 @@
+/**
+ * Heath & Fitness chat-plugin
+ * Pokemon Showdown - http://pokemonshowdown.com/
+ * 
```
trailing spaces........

Zarel commented on chat-plugins/health-fitness.js:
```diff
@@ -0,0 +1,168 @@
+ * Pokemon Showdown - http://pokemonshowdown.com/
+ * 
+ * This handles health and fitness daily challenges as well as a daily quote
+ * 
```
trailing spaces..........

Zarel commented:
iirc we removed whitespace checking on Windows because we were having trouble with Windows linebreaks...

But then why did Travis let this through? Travis tests on systems that aren't Windows, right???

@Slayer95 do you know what happened?

Slayer95 commented:
This commit happened https://github.com/Zarel/Pokemon-Showdown/commit/79a5b23f257f71988fd29805408e5711764c61ce

Slayer95 commented:
I mentioned the issue in review (https://github.com/Zarel/Pokemon-Showdown/pull/2350#issuecomment-168256648) but you still merged the PR...

Zarel commented:
You did? I don't see anything about trailing whitespace in your comments in #2350. Are we talking about different things? I'm talking about `eslint` checking for trailing whitespace.

Slayer95 commented:
Oh, trailing whitespace... I don't know about that.
That PR disables LF checks however.

Zarel commented:
Oh, yeah, because there's no way to force them in Windows in a `.eslintrc` file. Any solutions?
